### ISSUE TITLE
spookyhash segfaults with instances of reference classes

### ISSUE THREAD

**Issue by LTLA** [eddelbuettel/digest#146]
Pretty much as the title says; a minimal working example might be:

```r
example(setRefClass) # use this to generate 'xx' instance of reference class.
digest::digest(xx, algo="spookyhash")
## 
##  *** caught segfault ***
## address 0x5, cause 'memory not mapped'
## 
## Traceback:
##  1: paste(.Call(spookydigest_impl, object, skip, 0, 0, serializeVersion),     collapse = "")
##  2: digest::digest(xx, algo = "spookyhash")
## 
## Possible actions:
## 1: abort (with core dump, if enabled)
## 2: normal R exit
## 3: exit R without saving workspace
## 4: exit R saving workspace
```

I stumbled across this when trying to hash a larger object that happened to contain an instance of a reference class (namely a `SummarizedExperiment` instance, if you're curious).

I can reproduce this on the following systems:

<details>
<summary>Linux, R 3.6.1</summary>

```
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /XXX/lib64/R/lib/libRblas.so
LAPACK: /XXX/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] digest_0.6.23

loaded via a namespace (and not attached):
[1] compiler_3.6.1
```
</details>

<details>
<summary>MacOSX, R-devel</summary>

```
R Under development (unstable) (2019-11-04 r77366)
Platform: x86_64-apple-darwin17.7.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /XXX/lib/libRblas.dylib
LAPACK: /XXX/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] digest_0.6.23

loaded via a namespace (and not attached):
[1] compiler_4.0.0
```
</details>

Funnily enough, on R 3.6.1 on my Mac (same version of `digest`), I get:

```
Error in paste(.Call(spookydigest_impl, object, skip, 0, 0, serializeVersion),  :
  attempt to apply non-function
```

which is at least a bit nicer than a segfault.

**Comment by eddelbuettel** [comment]
I get what you get on Linux, R 3.6.2:

```r
R> digest::digest(xx)
[1] "9cc65858866993030fbf6e30ef119289"
R> digest::digest(xx, algo="spookyhash")
Error in paste(.Call(spookydigest_impl, object, skip, 0, 0, serializeVersion),  : 
  attempt to apply non-function
R> 
```

**Comment by eddelbuettel** [comment]
On rebuild I get the segfault.  It looks to me like me never serialized the object.

**Comment by eddelbuettel** [comment]
Paging @kendonB -- any chance you could look into this bug report on the spooky extension you contributed?

@LTLA We may just have to mark this as potentially buggy. Seems to copy with other data so...

**Comment by LTLA** [comment]
Ah, that's a shame. Thanks for the heads-up, I'll fall back to MD5.

**Comment by kendonB** [comment]
I have not solved this but I can narrow down a possible cause to between these two commits:

https://github.com/kendonB/digest/compare/1268b6a..ea043bb

The [base](https://github.com/kendonB/digest/commit/1268b6acc3a03aa36bf3073f779d9cadbc182776) is from [my first pull request that got rejected](https://github.com/eddelbuettel/digest/pull/98) and it seems to work:

```
> temp_library_location <- tempfile(pattern="lib")
> dir.create(temp_library_location)
> .libPaths(temp_library_location)
> .libPaths()
[1] "/tmp/Rtmply52FO/lib7b4a864c498" "/usr/local/lib/R/site-library" 
[3] "/usr/lib/R/site-library"        "/usr/lib/R/library"            
> 
> installed_before <- installed.packages()
> install.packages("https://github.com/kendonB/digest/archive/1268b6a.tar.gz", type = "source", repos = NULL, quiet = TRUE)
trying URL 'https://github.com/kendonB/digest/archive/1268b6a.tar.gz'
downloaded 120 KB

> installed_after <- installed.packages()
> 
> setdiff(installed_after[, "Package"], installed_before[, "Package"])
[1] "digest"
> 
> example(setRefClass, echo = FALSE) # use this to generate 'xx' instance of reference class.
Call:
$undo()


Undoes the last edit() operation
        and update the edits field accordingly.
        

Reference class 'mEdit' [package ".GlobalEnv"] with 2 fields
 $ data : num [1:4, 1:3] 1 2 3 4 5 6 7 8 9 10 ...
 $ edits: list()
 and 17 methods, of which 3 are  possibly relevant:
   edit, show#envRefClass, undo
> digest::digest(xx, algo = "spookyhash")
[1] "770d49a2b997d7f449e0f4bbab2e025d"
```
The [latter](https://github.com/kendonB/digest/commit/ea043bb2f7c53b4b83c4d159dadb1ccc83c766d1) is the first commit from the [PR that got merged](https://github.com/eddelbuettel/digest/pull/100) and it fails:
```
> temp_library_location <- tempfile(pattern="lib")
> dir.create(temp_library_location)
> .libPaths(temp_library_location)
> .libPaths()
[1] "/tmp/RtmpgnCN6B/lib7bcd1c1abfba" "/usr/local/lib/R/site-library"  
[3] "/usr/lib/R/site-library"         "/usr/lib/R/library"             
> 
> installed_before <- installed.packages()
> install.packages("https://github.com/kendonB/digest/archive/ea043bb.tar.gz", type = "source", repos = NULL, quiet = TRUE)
trying URL 'https://github.com/kendonB/digest/archive/ea043bb.tar.gz'
downloaded 120 KB

> installed_after <- installed.packages()
> 
> setdiff(installed_after[, "Package"], installed_before[, "Package"])
[1] "digest"
> 
> example(setRefClass, echo = FALSE) # use this to generate 'xx' instance of reference class.
Call:
$undo()


Undoes the last edit() operation
        and update the edits field accordingly.
        

Reference class 'mEdit' [package ".GlobalEnv"] with 2 fields
 $ data : num [1:4, 1:3] 1 2 3 4 5 6 7 8 9 10 ...
 $ edits: list()
 and 17 methods, of which 3 are  possibly relevant:
   edit, show#envRefClass, undo
> digest::digest("test", algo = "spookyhash")
[1] "a4f56ed6a46660d70f166f7c39f39029"
> digest::digest(xx, algo = "spookyhash")

 *** caught segfault ***
address 0x1, cause 'memory not mapped'

Traceback:
 1: paste(.Call(spooky_impl, object), collapse = "")
 2: digest::digest(xx, algo = "spookyhash")

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection: 
```
This diff fixes up the superfluous differences so you can see the changes that matter:
https://github.com/kendonB/digest/compare/717bc63..ea043bb Recall that the base works and the new commit fails.

The changes seem to do with registration, which I think I recall you helped me with @eddelbuettel before submitting the second PR. Can you see anything obvious that might be wrong?

Apologies for the lack of `reprex` reprexes, I couldn't get them to work while isolating the libPaths.

**Comment by eddelbuettel** [comment]
The registration business is harmless.  The one line that jumps out at me is the change from `R_xlen_t` to `uint32_t`.  That looks suspicious to me now.  R does not have unsigned ints natively, and these could possibly overflow (though probably hardly on an example data set).

Can you narrow down where the seg.fault happens in the C code?

**Comment by eddelbuettel** [comment]
And `digest.c` now has the `R_xlen_t` as it should.  So one guess may be that a cast goes wrong at your side?  Could that be?  It's really just a guess at this point.

Having a simpler test case may help too.

**Comment by kendonB** [comment]
I have tried making that minimal change:

https://github.com/kendonB/digest/compare/25fa1ca..ea043bb

The base here still works (and the revision fails) despite now the integer types being the same:

The base:
```r
temp_library_location <- tempfile(pattern="lib")
dir.create(temp_library_location)
.libPaths(temp_library_location)
install.packages("https://github.com/kendonB/digest/archive/25fa1ca.tar.gz", type = "source", repos = NULL, quiet = TRUE)
##> trying URL 'https://github.com/kendonB/digest/archive/25fa1ca.tar.gz'
downloaded 120 KB

##> > example(setRefClass, echo = FALSE) # use this to generate 'xx' instance of reference class.
##> Call:
##> $undo()


##> Undoes the last edit() operation
##>         and update the edits field accordingly.


##> Reference class 'mEdit' [package ".GlobalEnv"] with 2 fields
##>  $ data : num [1:4, 1:3] 1 2 3 4 5 6 7 8 9 10 ...
##>  $ edits: list()
##>  and 17 methods, of which 3 are  possibly relevant:
##>    edit, show#envRefClass, undo
# Works
digest::digest(xx, algo = "spookyhash")
##> [1] "89b4e2f62b9c938d9bafbd6e02141315"
``` 

The revision:
````r
temp_library_location <- tempfile(pattern="lib")
dir.create(temp_library_location)
.libPaths(temp_library_location)
install.packages("https://github.com/kendonB/digest/archive/ea043bb.tar.gz", type = "source", repos = NULL, quiet = TRUE)
trying URL 'https://github.com/kendonB/digest/archive/ea043bb.tar.gz'
downloaded 120 KB

example(setRefClass, echo = FALSE) # use this to generate 'xx' instance of reference class.
##> Call:
##> $undo()


##> Undoes the last edit() operation
##>         and update the edits field accordingly.
        

##> Reference class 'mEdit' [package ".GlobalEnv"] with 2 fields
##>  $ data : num [1:4, 1:3] 1 2 3 4 5 6 7 8 9 10 ...
##>  $ edits: list()
##>  and 17 methods, of which 3 are  possibly relevant:
##>    edit, show#envRefClass, undo
# Works
digest::digest("test", algo = "spookyhash")
##> [1] "a4f56ed6a46660d70f166f7c39f39029"
# Fails
digest::digest(xx, algo = "spookyhash")

##>  *** caught segfault ***
##> address 0x1, cause 'memory not mapped'

##> Traceback:
##>  1: paste(.Call(spooky_impl, object), collapse = "")
##>  2: digest::digest(xx, algo = "spookyhash")

##> Possible actions:
##> 1: abort (with core dump, if enabled)
##> 2: normal R exit
##> 3: exit R without saving workspace
##> 4: exit R saving workspace
##> Selection: 
```
I'm stumped.

**Comment by eddelbuettel** [comment]
I don't know either.  

But if we can't fix it now, maybe we need a line to the help page saying that it can occassionally barf to have this documented.   It's weird case, I still feel I am missing something here as it works otherwise -- we had the basic unit tests in since the start and they still pass.

**Comment by kendonB** [comment]
Is it an option to change the registration approach back to the way `fastdigest` had it? Maybe the original author @gmbecker has some insight?

**Comment by eddelbuettel** [comment]
No as that is against current CRAN Policy / best practices.  The registration save a look up, and just about every package does that now.  And `fastdigest` [does it too](https://github.com/gmbecker/fastdigest/blob/master/NAMESPACE#L1):

```r
useDynLib(fastdigest, .registration=TRUE)
```

**Comment by kendonB** [comment]
I believe this is fixed in #148.
